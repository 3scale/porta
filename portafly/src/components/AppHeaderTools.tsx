import React, { useState } from 'react'
import { useAuth } from 'auth'
import {
  Button,
  PageHeaderTools,
  PageHeaderToolsGroup,
  PageHeaderToolsItem,
  ButtonVariant,
  Dropdown,
  DropdownToggle,
  Avatar,
  KebabToggle,
  DropdownItem,
  DropdownGroup
} from '@patternfly/react-core'
import { useTranslation } from 'i18n/useTranslation'
import { CogIcon, HelpIcon } from '@patternfly/react-icons'
import avatar from 'assets/img_avatar.svg'

const AppHeaderTools: React.FunctionComponent = () => {
  const { t } = useTranslation('shared')
  const { userProfile, logout } = useAuth()

  const [isKebabDropdownOpen, setIsKebabDropdownOpen] = useState(false)
  const [isDropdownOpen, setIsDropDownOpen] = useState(false)

  const kebabDropdownItems = [
    <DropdownItem key="settings">
      <CogIcon />
      {' Settings'}
    </DropdownItem>,
    <DropdownItem key="help">
      <HelpIcon />
      {' Help'}
    </DropdownItem>
  ]

  const userDropdownItems = [
    <DropdownGroup key="group user">
      <DropdownItem key="group user logout" onClick={() => logout()}>
        {t('buttons.logout')}
      </DropdownItem>
    </DropdownGroup>
  ]

  return (
    <PageHeaderTools>
      <PageHeaderToolsGroup visibility={{ default: 'hidden', lg: 'visible' }}>
        <PageHeaderToolsItem>
          <Button aria-label="Settings actions" variant={ButtonVariant.plain}>
            <CogIcon />
          </Button>
        </PageHeaderToolsItem>
        <PageHeaderToolsItem>
          <Button aria-label="Help actions" variant={ButtonVariant.plain}>
            <HelpIcon />
          </Button>
        </PageHeaderToolsItem>
      </PageHeaderToolsGroup>
      <PageHeaderToolsGroup>
        <PageHeaderToolsItem visibility={{
          default: 'visible',
          sm: 'visible',
          md: 'visible',
          lg: 'hidden',
          xl: 'hidden',
          '2xl': 'hidden'
        }}
        >
          <Dropdown
            isPlain
            position="right"
            onSelect={() => setIsKebabDropdownOpen(!isKebabDropdownOpen)}
            toggle={<KebabToggle onToggle={() => setIsKebabDropdownOpen(!isKebabDropdownOpen)} />}
            isOpen={isKebabDropdownOpen}
            dropdownItems={kebabDropdownItems}
          />
        </PageHeaderToolsItem>
        <PageHeaderToolsItem visibility={{
          default: 'hidden',
          sm: 'hidden',
          md: 'hidden',
          lg: 'visible',
          xl: 'visible',
          '2xl': 'visible'
        }}
        >
          <Dropdown
            isPlain
            position="right"
            onSelect={() => setIsDropDownOpen(!isDropdownOpen)}
            isOpen={isDropdownOpen}
            toggle={(
              <DropdownToggle onToggle={() => setIsDropDownOpen(!isDropdownOpen)}>
                {`${userProfile.firstName} ${userProfile.lastName}`}
              </DropdownToggle>
              )}
            dropdownItems={userDropdownItems}
          />
        </PageHeaderToolsItem>
      </PageHeaderToolsGroup>
      <Avatar src={avatar} alt="Avatar image" />
    </PageHeaderTools>
  )
}

export { AppHeaderTools }
