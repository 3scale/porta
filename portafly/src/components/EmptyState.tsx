import React from 'react'
import {
  Title,
  EmptyState,
  EmptyStateVariant,
  EmptyStateIcon,
  Bullseye,
  EmptyStateBody
} from '@patternfly/react-core'
import { CubesIcon, SearchIcon } from '@patternfly/react-icons'

export interface ISimpleEmptyStateProps {
  msg: string
}

const SimpleEmptyState: React.FunctionComponent<ISimpleEmptyStateProps> = ({ msg }) => (
  <EmptyState variant={EmptyStateVariant.full}>
    <EmptyStateIcon icon={CubesIcon} />
    <Title headingLevel="h5" size="lg">
      {msg}
    </Title>
  </EmptyState>
)

interface ITableEmptyState {
  title: string
  body: string
  button?: JSX.Element
}

const TableEmptyState: React.FunctionComponent<ITableEmptyState> = ({
  title,
  body,
  button
}) => (
  <Bullseye>
    <EmptyState variant={EmptyStateVariant.small}>
      <EmptyStateIcon icon={SearchIcon} />
      <Title headingLevel="h2" size="lg">
        {title}
      </Title>
      <EmptyStateBody>
        {body}
      </EmptyStateBody>
      {button}
    </EmptyState>
  </Bullseye>
)

export { SimpleEmptyState, TableEmptyState }
