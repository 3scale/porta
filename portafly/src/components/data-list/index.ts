export * from './BulkSelectorWidget'
export * from './BulkActionsWidget'
export * from './DataListContext'
export * from './SearchWidget'
export * from './Toolbar'
export * from './PaginationWidget'
export * from './modals'
export * from './utils'
