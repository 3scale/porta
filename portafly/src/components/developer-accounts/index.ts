export * from './ActionsDropdown'
export * from './BulkActionSelector'
export * from './SearchWidget'
export * from './DeveloperAccountsTable'
export * from './ActionButtonImpersonate'
export * from './util/tableDataFactory'
