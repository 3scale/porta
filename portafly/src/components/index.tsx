export * from './pages'
export * from './shared'
export * from './AppLayout'
export * from './AppNavExpandable'
export * from './AppNavGroup'
export * from './AppNavItem'
export * from './AppHeaderTools'
export * from './LazyRoute'
export * from './NotFound'
export * from './SwitchWith404'
export * from './util'
export * from './Root'
