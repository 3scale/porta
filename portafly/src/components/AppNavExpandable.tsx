import * as React from 'react'
import { useRouteMatch } from 'react-router-dom'
import { NavExpandable } from '@patternfly/react-core'
import { AppNavItem, IAppNavItemProps } from 'components/AppNavItem'

export interface IAppNavExpandableProps {
  title: string
  to: string
  items: Array<IAppNavItemProps | undefined>
}

export const AppNavExpandable: React.FunctionComponent<IAppNavExpandableProps> = ({
  title,
  to,
  items
}) => {
  const match = useRouteMatch({
    path: to
  })
  const isActive = !!match
  return (
    <NavExpandable title={title} isActive={isActive} isExpanded={isActive}>
      {items.map((item, j) => (
        // FIXME: each app-nav-item should have a fixed ID, using title? or j is just a dirty trick
        <AppNavItem title={item?.title} to={item?.to} key={item?.title || j} />
      ))}
    </NavExpandable>
  )
}
