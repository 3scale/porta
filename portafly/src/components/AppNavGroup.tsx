import * as React from 'react'
import { NavGroup } from '@patternfly/react-core'
import { AppNavItem, IAppNavItemProps } from 'components/AppNavItem'

export interface IAppNavGroupProps {
  title: string
  items: Array<IAppNavItemProps | undefined>
}

export const AppNavGroup: React.FunctionComponent<IAppNavGroupProps> = ({
  title,
  items
}) => (
  <NavGroup title={title}>
    {items.map((i) => (
      <AppNavItem title={i?.title} to={i?.to} exact={i?.exact} key={i?.title} />
    ))}
  </NavGroup>
)
