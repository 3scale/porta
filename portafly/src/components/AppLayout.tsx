import * as React from 'react'
import {
  Nav,
  NavList,
  Page,
  PageHeader,
  PageSidebar,
  SkipToContent,
  PageHeaderProps
} from '@patternfly/react-core'
import { AppNavExpandable, IAppNavExpandableProps } from 'components/AppNavExpandable'
import { AppNavGroup, IAppNavGroupProps } from 'components/AppNavGroup'
import { AppNavItem, IAppNavItemProps } from 'components/AppNavItem'

export interface IAppLayoutContext {
  setBreadcrumb: (breadcrumb: React.ReactNode) => void
}

export const AppLayoutContext = React.createContext<IAppLayoutContext>({
  setBreadcrumb: () => 0
})

export interface IAppLayoutProps extends Pick<PageHeaderProps, 'logo' | 'logoProps' | 'headerTools'> {
  navVariant?: 'vertical' | 'horizontal'
  navItems?: Array<IAppNavItemProps | IAppNavExpandableProps | IAppNavGroupProps | undefined>
  navGroupsStyle?: 'grouped' | 'expandable'
  startWithOpenNav?: boolean
  theme?: 'dark' | 'light'
  mainContainerId?: string
}

export const AppLayout: React.FunctionComponent<IAppLayoutProps> = ({
  logo,
  logoProps,
  navVariant = 'horizontal',
  navItems = [],
  navGroupsStyle = 'grouped',
  headerTools,
  startWithOpenNav = true,
  theme = 'dark',
  mainContainerId = 'main-container',
  children
}) => {
  const [isNavOpen, setIsNavOpen] = React.useState(startWithOpenNav)
  const [isMobileView, setIsMobileView] = React.useState(true)
  const [isNavOpenMobile, setIsNavOpenMobile] = React.useState(false)
  const [breadcrumb, setBreadcrumb] = React.useState<React.ReactNode | undefined>()
  const previousBreadcrumb = React.useRef<React.ReactNode | null>()

  const handleSetBreadcrumb = React.useCallback(
    (newBreadcrumb: React.ReactNode) => {
      if (previousBreadcrumb.current !== newBreadcrumb) {
        previousBreadcrumb.current = newBreadcrumb
        setBreadcrumb(previousBreadcrumb.current)
      }
    },
    [setBreadcrumb, previousBreadcrumb]
  )

  const onNavToggleMobile = React.useCallback(() => {
    setIsNavOpenMobile(!isNavOpenMobile)
  }, [setIsNavOpenMobile, isNavOpenMobile])

  const onNavToggle = React.useCallback(() => {
    setIsNavOpen(!isNavOpen)
  }, [setIsNavOpen, isNavOpen])

  const onPageResize = (props: { mobileView: boolean; windowSize: number }) => {
    setIsMobileView(props.mobileView)
  }

  React.useEffect(() => {
    setIsNavOpen(startWithOpenNav)
  }, [startWithOpenNav, setIsNavOpen])

  const isVertical = navVariant === 'vertical'
  const variant = isVertical ? 'default' : 'horizontal'

  const Navigation = React.useMemo(
    () => navItems.length > 0 && (
    <Nav id="nav-primary-simple" theme={theme} variant={variant}>
      <NavList id="nav-list-simple">
        {navItems.map((navItem) => {
          // eslint-disable-next-line no-prototype-builtins
          if (navItem && navItem.hasOwnProperty('items') && isVertical) {
            if (navGroupsStyle === 'expandable') {
              const { title, to, items } = navItem as IAppNavExpandableProps
              return <AppNavExpandable title={title} to={to} items={items} key={title} />
            }

            const { title, items } = navItem as IAppNavGroupProps
            return <AppNavGroup title={title} items={items} key={title} />
          }

          // TODO: extract NavItemSeparator from AppNavItem to clean this up
          const item = navItem as (IAppNavItemProps | undefined)
          return (
            <AppNavItem
              title={item?.title}
              to={item?.to}
              exact={item?.exact}
              key={item?.title}
            />
          )
        })}
      </NavList>
    </Nav>
    ),
    [isVertical, navGroupsStyle, navItems, theme, variant]
  )

  const Header = React.useMemo(
    () => (
      <PageHeader
        logo={logo}
        logoProps={logoProps}
        headerTools={headerTools}
        showNavToggle={isVertical}
        isNavOpen={isVertical && isNavOpen}
        onNavToggle={isMobileView ? onNavToggleMobile : onNavToggle}
        topNav={isVertical ? undefined : Navigation}
      />
    ),
    [
      logo,
      logoProps,
      headerTools,
      isVertical,
      isNavOpen,
      isMobileView,
      onNavToggle,
      onNavToggleMobile,
      Navigation
    ]
  )

  const Sidebar = React.useMemo(
    () => navVariant === 'vertical' && (
    <PageSidebar
      nav={Navigation}
      isNavOpen={isMobileView ? isNavOpenMobile : isNavOpen}
      theme={theme}
      data-testid="app-sidebar"
    />
    ),
    [navVariant, Navigation, isMobileView, isNavOpenMobile, isNavOpen, theme]
  )
  const PageSkipToContent = (
    <SkipToContent href="#primary-app-container">
      Skip to Content
    </SkipToContent>
  )
  return (
    <AppLayoutContext.Provider value={{ setBreadcrumb: handleSetBreadcrumb }}>
      <Page
        mainContainerId={mainContainerId}
        header={Header}
        sidebar={Sidebar}
        breadcrumb={breadcrumb}
        onPageResize={onPageResize}
        skipToContent={PageSkipToContent}
      >
        {children}
      </Page>
    </AppLayoutContext.Provider>
  )
}
