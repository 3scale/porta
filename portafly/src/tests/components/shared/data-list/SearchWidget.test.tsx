import React from 'react'

import { render } from 'tests/custom-render'
import { fireEvent } from '@testing-library/react'
import { SearchWidget } from 'components'
import { Toolbar, ToolbarContent, ToolbarItem } from '@patternfly/react-core'

const categories = [
  {
    name: 'admin',
    humanName: 'Admin'
  },
  {
    name: 'group',
    humanName: 'Organization / Group'
  },
  {
    name: 'state',
    humanName: 'State',
    options: [
      {
        name: 'active',
        humanName: 'Active'
      },
      {
        name: 'pending',
        humanName: 'Pending'
      }
    ]
  }
]

describe('SearchWidget', () => {
  const setup = () => (
    render(
      <Toolbar id="data-toolbar">
        <ToolbarContent>
          <ToolbarItem data-testid="search-widget">
            <SearchWidget categories={categories} />
          </ToolbarItem>
        </ToolbarContent>
      </Toolbar>
    )
  )

  it('should render correctly', () => {
    const { getByTestId } = setup()
    expect(getByTestId('search-widget')).toMatchSnapshot()
  })

  it.skip('should be able to filter by admin', () => {
    // FIXME: Update test after MOB session
    const { getByPlaceholderText, getByTestId } = setup()
    const searchInput = getByPlaceholderText('accountsIndex:accounts_filter_field')
    fireEvent.change(searchInput, { target: { value: 'big boss' } })
    expect(getByTestId('data-toolbar')).toMatchSnapshot()
  })

  // TODO: PF React makes it hard to target the button of the select...
  it.skip('should be able to filter by state', () => {
  })
})
