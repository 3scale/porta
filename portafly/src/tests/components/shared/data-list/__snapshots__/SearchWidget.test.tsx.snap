// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchWidget should be able to filter by admin 1`] = `
<div
  class="pf-c-toolbar__item"
  data-testid="search-widget"
>
  <div
    class="pf-c-toolbar__group pf-m-filter-group"
  >
    <div
      class="pf-c-select"
      data-ouia-component-id="0"
      data-ouia-component-type="PF4/Select"
      data-ouia-safe="true"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby=" pf-toggle-id-0"
        class="pf-c-select__toggle"
        id="pf-toggle-id-0"
        type="button"
      >
        <div
          class="pf-c-select__toggle-wrapper"
        >
          <span
            class="pf-c-select__toggle-icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                transform=""
              />
            </svg>
          </span>
          <span
            class="pf-c-select__toggle-text"
          >
            Admin
          </span>
        </div>
        <span
          class="pf-c-select__toggle-arrow"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="pf-c-toolbar__item"
    >
      <span />
    </div>
    <div
      class="pf-c-toolbar__item"
    >
      <span />
    </div>
    <div
      class="pf-c-toolbar__item"
    >
      <span />
    </div>
    <div
      class="pf-c-input-group"
    >
      <input
        aria-invalid="false"
        aria-label="accountsIndex:accounts_filter_field_aria_label"
        class="pf-c-form-control"
        placeholder="accountsIndex:accounts_filter_field"
        type="search"
        value=""
      />
      <button
        aria-disabled="false"
        class="pf-c-button pf-m-control"
        data-ouia-component-id="1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            transform=""
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`SearchWidget should render correctly 1`] = `
<div
  class="pf-c-toolbar__item"
  data-testid="search-widget"
>
  <div
    class="pf-c-toolbar__group pf-m-filter-group"
  >
    <div
      class="pf-c-select"
      data-ouia-component-id="0"
      data-ouia-component-type="PF4/Select"
      data-ouia-safe="true"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby=" pf-toggle-id-0"
        class="pf-c-select__toggle"
        id="pf-toggle-id-0"
        type="button"
      >
        <div
          class="pf-c-select__toggle-wrapper"
        >
          <span
            class="pf-c-select__toggle-icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                transform=""
              />
            </svg>
          </span>
          <span
            class="pf-c-select__toggle-text"
          >
            Admin
          </span>
        </div>
        <span
          class="pf-c-select__toggle-arrow"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="pf-c-toolbar__item"
    >
      <span />
    </div>
    <div
      class="pf-c-toolbar__item"
    >
      <span />
    </div>
    <div
      class="pf-c-toolbar__item"
    >
      <span />
    </div>
    <div
      class="pf-c-input-group"
    >
      <input
        aria-invalid="false"
        aria-label="accountsIndex:accounts_filter_field_aria_label"
        class="pf-c-form-control"
        placeholder="accountsIndex:accounts_filter_field"
        type="search"
        value=""
      />
      <button
        aria-disabled="false"
        aria-label="accountsIndex:accounts_filter_button_aria_label"
        class="pf-c-button pf-m-control"
        data-ouia-component-id="1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            transform=""
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
